src/rollup/rollup.ts

```diff
@@ -56,6 +56,7 @@ export async function rollupInternal(
 			err.watchFiles = watchFiles;
 		}
 		await graph.pluginDriver.hookParallel('buildEnd', [err]);
+		if (!watcher) await graph.pluginDriver.hookParallel('closeBundle', []);
```

Comment:
Good catch doing this on errors. I am wondering very much about how to handle watch mode, though. My feeling is at the moment that it is easiest to reason about this feature as "closing a bundle means no more generate/write calls will happen for this specific bundle". This would mean that even in watch mode, we should call "closeBundle" after the call to bundle.write.

This means basically:
- Always call `closeBundle` here even if there is a watcher
- Add a call `await bundle.close()` in `watch.ts`, e.g. after line 210 (i.e. after the calls to `result.write`).

This would have some implications for plugins, though:
- When `bundleClose` is called, it does not mean the plugin is definitely done. They could still clean up their resources, but they should be ready to spin them up again if another call to `buildStart` follows.
- Alternatively, they can check for [`this.meta.watchMode`](https://rollupjs.org/guide/en/#thismeta-rollupversion-string-watchmode-boolean) to behave differently in watch mode. But then they would need to find a different way for cleanup such as attaching a `process.on('SIGINT', ...)` listener.

Reply:
Ah, maybe I am wrong about the point of adding `await bundle.close()` in `watch.ts L210`. Because as we are exposing the result, users could still call `generate` as often as they want. But I think we should call `closeBundle` unconditionally on errors.

As for successful builds in watch mode, I guess the task of closing the bundle would then fall to the handler for the `BUNDLE_END` event in `watch-cli.ts`. And we should document for watch mode that it is the task of whoever uses the JavaScript API for watch mode, they should call `await result.close()` in response to the `BUNDLE_END` event once done.

Reply:
Ah, so watch mode continues to run `closeBundle` on failure, but as for successful builds, users would have to be the one closing it via `BUNDLE_END`?

Have to admit I forgot that the watcher is exposed on the API, so I didn't pay much attention to this. One thing I'm concerned is that it would be confusing to explain on how to close the bundles properly.

```js
let watcher = rollup.watch({ ... });
watcher.on('event', (event) => { ... });

// "Shouldn't this close the bundles for you?"
watcher.close();
```

It would then lead to users unnecessarily calling `result.close()` each time a bundle is finished building, which adds overhead to plugins that need to start an external service/process in order to do something, so one way for users to close it down properly would be to do something like this

```js
let bundles = [];
let watcher = rollup.watch({ ... });

watcher.on('event', (event) => {
  switch (event.code) {
    case 'START':
      bundles.length = 0;
      break;
    case 'BUNDLE_END':
      bundles.push(event.result);
      break;
  }
});

watcher.close();
bundles.forEach((bundle) => bundle.close());
```

Reply:
> Ah, so watch mode continues to run closeBundle on failure, but as for successful builds, users would have to be the one closing it via BUNDLE_END

Yes. The reason they do not have to close the bundle manually for unsuccessful builds is because, well, there is no bundle for unsuccessful build! And there is also no `BUNDLE_END` event. So even if they wanted to, they could not do it manually.

> which adds overhead to plugins that need to start an external service/process in order to do something

If plugins do not specifically detect watch mode, yes. But if you do NOT run the hook after each run in watch mode, plugins that WOULD want to do some cleanup after each run have no chance to detect the current run is complete. No, I do not think we should mix concerns here. Also, your proposal would quickly fill up the memory with all bundles created without removing any, so basically a guaranteed memory leak given some time running watch mode.

No, if plugins want to retain state across runs in watch mode, then they should specifically check if they are in watch mode via `this.meta.watchMode` and then not perform cleanup in watch mode.

Reply:
Ahh, okay, understood